lib/SILOptimizer/Analysis/ProtocolConformanceAnalysis.cpp

```diff
@@ -0,0 +1,70 @@
+//===-- ProtocolConformanceAnalysis.cpp - Protocol Conformance Analysis ---===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+// This analysis collects a set of nominal types (classes, structs, and enums)
+// that conform to a protocol during whole module compilation. We only track
+// protocols that are non-public.
+
+#include "swift/SILOptimizer/Analysis/ProtocolConformanceAnalysis.h"
+#include "swift/AST/ASTContext.h"
+#include "swift/AST/ASTWalker.h"
+#include "swift/AST/Module.h"
+#include "swift/SIL/SILInstruction.h"
+#include "swift/SIL/SILModule.h"
+#include "swift/SIL/SILValue.h"
+
+using namespace swift;
+
+namespace {
+/// A helper class to collect all nominal type declarations.
+class NominalTypeWalker : public ASTWalker {
+  ProtocolConformanceAnalysis::ProtocolConformanceMap &ProtocolConformanceCache;
+
+public:
+  NominalTypeWalker(ProtocolConformanceAnalysis::ProtocolConformanceMap
+                    &ProtocolConformanceCache)
+      : ProtocolConformanceCache(ProtocolConformanceCache) {}
+
+  bool walkToDeclPre(Decl *D) override {
+    if (auto *NTD = dyn_cast<NominalTypeDecl>(D)) {
+      if (NTD->getEffectiveAccess() <= AccessLevel::Internal) {
```

Comment:
Isn't the right check here for an internal _protocol_, not an internal conforming type? Public protocols can always have conforming types from outside the module.